src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressWithNearbyTextFilter.java

```diff
@@ -0,0 +1,620 @@
+    public final void setCheckPattern(String pattern) {
+        checkPattern = pattern;
+    }
+
+    /**
+     * Setter to specify check violation message pattern to suppress.
+     *
+     * @param pattern a {@code String} value.
+     */
+    public void setMessagePattern(String pattern) {
+        messagePattern = pattern;
+    }
+
+    /**
+     * Setter to specify check ID pattern to suppress.
+     *
+     * @param pattern a {@code String} value.
+     */
+    public void setIdPattern(String pattern) {
+        idPattern = pattern;
+    }
+
+    /**
+     * Setter to specify negative/zero/positive value that defines the number
+     * of lines preceding/at/following the suppressing nearby text. Property can also
+     * be a RegExp group index at {@code nearbyTextPattern} in
+     * format of {@code $x} and be picked from line that matches {@code nearbyTextPattern}.
+     *
+     * @param format a {@code String} value.
+     */
+    public final void setLineRange(String format) {
+        lineRange = format;
+    }
+
+    @Override
+    public boolean accept(AuditEvent event) {
+        boolean accepted = true;
+
+        if (event.getViolation() != null) {
+            final String eventFileTextAbsolutePath = event.getFileName();
+
+            if (!currentFileAbsolutePath.equals(eventFileTextAbsolutePath)) {
+                final FileText currentFileText = getFileText(eventFileTextAbsolutePath);
+
+                if (currentFileText != null) {
+                    currentFileAbsolutePath = currentFileText.getFile().getAbsolutePath();
+                    collectSuppressions(currentFileText);
+                }
+            }
+
+            final Optional<Suppression> nearestSuppression =
+                    getNearestSuppression(suppressions, event);
+            accepted = nearestSuppression.isEmpty();
+        }
+        return accepted;
+    }
+
+    @Override
+    protected void finishLocalSetup() {
+        // No code by default
+    }
+
+    /**
+     * Returns {@link FileText} instance created based on the given file name.
+     *
+     * @param fileName the name of the file.
+     * @return {@link FileText} instance.
+     * @throws IllegalStateException if the file could not be read.
+     */
+    private static FileText getFileText(String fileName) {
+        final File file = new File(fileName);
+        FileText result = null;
+
+        if (!file.isDirectory()) {
+            try {
+                result = new FileText(file, StandardCharsets.UTF_8.name());
+            }
+            catch (IOException ex) {
+                throw new IllegalStateException("Cannot read source file: " + fileName, ex);
+            }
+        }
+
+        return result;
+    }
+
+    /**
+     * Collets all {@link Suppression} instances retrieved from the given {@link FileText}.
+     *
+     * @param fileText {@link FileText} instance.
+     */
+    private void collectSuppressions(FileText fileText) {
+        suppressions.clear();
```

Comment:
is it possible to make suppressions not a field ? we need to strive to state-less and immutable state of Filter.

Reply:
I cannot think how we can make it without a field. Collecting the suppressions on every `AuditEvent` would be inefficient.

Reply:
Is it same solution in other filters?

Reply:
50/50
`SuppressWithNearbyCommentFilter` does keep them in a field:
https://github.com/checkstyle/checkstyle/blob/49b83057eb2265a25b976361b0fdb0f4f64c2c44/src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressWithNearbyCommentFilter.java#L276-L277

`SuppressWithPlainTextCommentFilter` doesn't have a field and gets suppressions on every audit event:
https://github.com/checkstyle/checkstyle/blob/49b83057eb2265a25b976361b0fdb0f4f64c2c44/src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressWithPlainTextCommentFilter.java#L393-L400

Reply:
ok, lets keep it, we can improve on this later on, as we come to MTthread mode of execution.